String format nonliteral warnings

When compiling with the flag `-Wformat-nonliteral`, the following warnings are produced by Clang:

```cpp
include/NAS2D/StringUtils.h:41:61: warning: format string is not a string literal [-Wformat-nonliteral]
                std::size_t size = snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
                                                                          ^~~~~~~~~~~~~~
include/NAS2D/StringUtils.h:43:42: warning: format string is not a string literal [-Wformat-nonliteral]
                snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
                                                       ^~~~~~~~~~~~~~
```

The warning is meant to scan for potential safety issues with string format functions. As the parameter list is fixed at compile time, the format specifier types must also be fixed at compile time to match, otherwise you get safety issues. Effectively, the only part of a format string that should change at runtime is the text between specifiers, or format options (but not the type). Even the order is fixed.

For the affected code, all of them use literal strings for the format, but because of the template function wrapper, the compiler doesn't see that.

The warning is correct in that it is possible to call the function with a non-literal string, which would have serious safety implications.

----

One potential solution might have been to move the format string to a template non-type parameter. Instead of `string_format("%d", 0)`, it would become `string_format<"%d">(0)`. However, this doesn't work because string literals have internal linkage, and such template non-type parameters need external linkage.

```
error: no matching function for call to 'string_format'
...
note: candidate template ignored: invalid explicitly-specified argument for template parameter 'format'
```

Reference: [String literals not allowed as non type template parameters](https://stackoverflow.com/questions/5547852/string-literals-not-allowed-as-non-type-template-parameters)

A possible workaround is to declare a named string constant, so that it can be given external linkage, however that's a sufficient pain so as to not be practical.

----

Another solution is to use [`std::to_string`](https://en.cppreference.com/w/cpp/string/basic_string/to_string) instead of the custom `string_format` method.

All current uses of the `string_format` template function are basically to format integers with the format specifier `"%i"`. This is easily accomplished by using `std::to_string`.

The first post contained the error message from Clang. A similar error message is produced by GCC-8 when compiling with `-Wformat-nonliteral`:
```
include/NAS2D/StringUtils.h:41:30: warning: format not a string literal, argument types not checked [-Wformat-nonliteral]
   std::size_t size = snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
                      ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
include/NAS2D/StringUtils.h:43:11: warning: format not a string literal, argument types not checked [-Wformat-nonliteral]
   snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I recommend removing this method and the associated unit test code.

--- a/include/NAS2D/StringUtils.h
+++ b/include/NAS2D/StringUtils.h
@@ -34,17 +34,6 @@ namespace NAS2D
 	bool startsWith(std::string_view string, char start) noexcept;
 	bool endsWith(std::string_view string, char end) noexcept;
 
-	template<typename... Args>
-	std::string string_format(const std::string& format, Args... args)
-	{
-		std::string buffer;
-		std::size_t size = snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
-		buffer.resize(size + 1); // Including null (avoid clipping/undefined behavior)
-		snprintf(buffer.data(), buffer.size(), format.c_str(), args...);
-		buffer.resize(size); // Strip null temrinator
-		return buffer;
-	}
-
 	/**
  * \typedef StringList
  * \brief	A list of std::string's.

--- a/test/StringUtils.test.cpp
+++ b/test/StringUtils.test.cpp
@@ -36,26 +36,3 @@ TEST(String, splitSkipEmpty)
 	EXPECT_EQ((NAS2D::StringList{"ab", "c"}), NAS2D::splitSkipEmpty("ab.c", '.'));
 	EXPECT_EQ((NAS2D::StringList{"abc"}), NAS2D::splitSkipEmpty("abc.", '.'));
 }
-
-TEST(String, string_format)
-{
-	// Numbers
-	EXPECT_EQ("-1", NAS2D::string_format("%d", -1));
-	EXPECT_EQ("0", NAS2D::string_format("%d", 0));
-	EXPECT_EQ("1", NAS2D::string_format("%d", 1));
-
-	// Padded numbers
-	EXPECT_EQ(" 10", NAS2D::string_format("%3d", 10));
-	EXPECT_EQ("010", NAS2D::string_format("%03d", 10));
-	// Negative padded numbers
-	EXPECT_EQ("-10", NAS2D::string_format("%3d", -10));
-	EXPECT_EQ("-10", NAS2D::string_format("%03d", -10));
-	EXPECT_EQ(" -10", NAS2D::string_format("%4d", -10));
-	EXPECT_EQ("-010", NAS2D::string_format("%04d", -10));
-
-	// Strings
-	EXPECT_EQ("Hello World", NAS2D::string_format("Hello %s", "World"));
-
-	// Padded strings
-	EXPECT_EQ("  ABC", NAS2D::string_format("%5s", "ABC"));
-}

